SU2_CFD/src/solver_direct_mean_inc.cpp

```diff
@@ -5308,57 +5239,64 @@ void CIncEulerSolver::BC_Sym_Plane(CGeometry      *geometry,
-        else                                              {Largest=2;Arbitrary=1;Zero=0;}
-
-        Tangential[Largest] = -UnitNormal[Arbitrary]/sqrt(pow(UnitNormal[Largest],2) + pow(UnitNormal[Arbitrary],2));
-        Tangential[Arbitrary] =  UnitNormal[Largest]/sqrt(pow(UnitNormal[Largest],2) + pow(UnitNormal[Arbitrary],2));
-        Tangential[Zero] =  0.0;
-        break;
-      }
-    }
-  }
   
   /*--- Loop over all the vertices on this boundary marker. ---*/
   for (iVertex = 0; iVertex < geometry->nVertex[val_marker]; iVertex++) {
     
+    /*---------------------------------------------------------------------------------------------*/
+    /*--- Preprocessing: On a symmetry-plane, the Unit-Normal is constant. Therefore a constant ---*/
+    /*---                Unit-Tangential to that Unit-Normal can be prescribed. The computation ---*/
+    /*---                of these vectors is done outside the loop (over all Marker-vertices).  ---*/
+    /*---                The "Normal" in SU2 is an Area-Normal and is most likely not constant  ---*/
+    /*---                on the symmetry-plane.                                                 ---*/
+    /*--- Edit July 2019: In order to use this BC_Sym_Plane method for slip walls in viscous    ---*/
+    /*---                 flow the unit-normal & tangent computation is moved into the loop over---*/
+    /*---                 all vertices. Slip walls can have a non-straight line or plane as a   ---*/
+    /*---                 boundary, therefore the assumption of a constant unit-normal is not   ---*/
+    /*---                 valid in these cases. https://github.com/su2code/SU2/issues/735       ---*/
+    /*---------------------------------------------------------------------------------------------*/
+
+    /*--- Normal vector for a random vertex (zero) on this marker (negate for outward convention). ---*/
+    if (iVertex == 0 || !(geometry->bound_is_straight[val_marker]) ) {
+      geometry->vertex[val_marker][iVertex]->GetNormal(Normal); 
+      for (iDim = 0; iDim < nDim; iDim++)
+        Normal[iDim] = -Normal[iDim];
+
+      /*--- Compute unit normal, to be used for unit tangential, projected velocity and velocity component gradients. ---*/
+      Area = 0.0;
+      for (iDim = 0; iDim < nDim; iDim++)
+        Area += Normal[iDim]*Normal[iDim];
+      Area = sqrt (Area);
+
+      for (iDim = 0; iDim < nDim; iDim++)
+        UnitNormal[iDim] = -Normal[iDim]/Area;
+    
+      /*--- Preprocessing: Compute unit tangential, the direction is arbitrary as long as t*n=0 ---*/
+      if (config->GetViscous()) {
+        switch( nDim ) {
+          case 2: {
+            Tangential[0] = -UnitNormal[1];
+            Tangential[1] =  UnitNormal[0];
+            break;
+          }
+          case 3: {
+            /*--- Find the largest entry index of the UnitNormal, and create Tangential vector based on that. ---*/
+            unsigned short Largest, Arbitrary, Zero;
```

Comment:
I think the method in the top voted answer should allow you to do this with less logic.
https://math.stackexchange.com/questions/137362/how-to-find-perpendicular-vector-to-another-vector

Reply:
Well, 1. It has to be a unit normal, therefore the division by sqrt(...) and 2. this Largest, Arbitratry, Zero stuff (which I understand bothers you) is there to avoid having the unit vector build upon numerical zeros which I found to give wrong results. Example: normal is n=( 1e-13, 1e-13, 1)  then I could just take t=(n2, -n1, 0) (which is in fact the way v_1 from that top voted answer is build). Then t is no unit vector so I divide each entry by the norm of t which ends up with inaccurate tangentials. That problem occurred in my tests for #657 . I hope that makes it clearer why it is like that. Search the largest entry of the vector and do that switch (like here first and second  value t=(n2, -n1, 0)  with that largest value

Or did I misinterpret your suggestion

Reply:
Yeah I got the purpose, but he does a linear combination of two trivially perpendicular vectors.
Which, if the coefficients of the combination are arbitrary, can still give you a small vector.
So one has to choose the coefficients that maximize the magnitude of the potentially zero element (that's one "if").
Worst case scenario this keeps the largest and smallest element in the tangent vector, but if one adds a third trivially perpendicular vector to the result that is avoided:
```
n = ai + bj +ck
if |b| > |c|
  t = bi + (c-a)j - bk
else
  t = ci - cj + (b-a)k
```
Which can actually be written without the conditional. But anyway maybe this is not even expensive at all, I just puzzles.

Reply:
Ok, got it. And I agree that it is the more elegant solution so I took it.